purchase_post_code_propose_address_sst/models/purchase_order.py

```diff
@@ -0,0 +1,115 @@
+    _inherit = 'purchase.order'
+
+    state_id = fields.Many2one(
+        'res.country.state',
+        string='Prefectures',
+    )
+    city = fields.Char(
+        string='City',
+    )
+    street = fields.Char(
+        oldname='address',
+        string='Street',
+    )
+    zipcode= fields.Char(
+        string='Post Code (Search)',
+    )
+    zip = fields.Char(
+        related='partner_id.zip',
+        string='Post Code (Supplier)',
+        store=True,
+        readonly=True,
+    )
+
+    @api.onchange('partner_id')
+    def _onchange_partner_id(self):
+        if self.partner_id:
+            self.state_id = self.partner_id.state_id
+            self.city = self.partner_id.city
+            self.street = self.partner_id.street
+
+    @api.onchange('zipcode')
+    def _onchange_zipcode(self):
+        if self.zipcode:
+            self.zipcode, msg = self.check_zipcode(self.zipcode)
+            if not self.zipcode:
+                return msg
+            request_url = 'http://zipcloud.ibsnet.co.jp/api/search?zipcode' \
+                          '=%s' % self.zipcode
+            request = urllib.request.Request(request_url)
+            response_data = json.loads(urllib.request.urlopen(request).read())
+            self.state_id = False
+            self.city = False
+            self.street = False
+            if response_data['status'] != 200:
+                self.zipcode = False
+                return {
+                    'warning': {
+                        'title': _("Error"),
+                        'message': response_data['message']
+                    }
+                }
+            else:
+                address_data = response_data['results']
+                if address_data:
+                    self.state_id = self.env['res.country.state'].search([
+                        ('name', '=', address_data[0]['address1'])
+                    ], limit=1)
+                    self.city = address_data[0]['address2']
+                    self.street = address_data[0]['address3']
+
+    def check_zipcode(self, zipcode):
+        msg = {}
+        field = jaconv.z2h(zipcode, ascii=True, digit=True).replace("-", "")
+        if not field.isdigit():
+            field = False
+            msg = {
+                'warning': {
+                    'title': _("Error"),
+                    'message': _("Only digits are allowed.")
+                }
+            }
+        if len(field) != 7:
+            field = False
+            msg = {
+                'warning': {
+                    'title': _("Error"),
+                    'message': _("Post code should be 7 digits.")
+                }
+            }
+        return field, msg
+
+    @api.multi
+    def write(self, vals):
+        res = super(PurchaseOrder, self).write(vals)
+        for order in self:
+            if not self.is_default_partner(order.partner_id.id) and \
+                    order.zipcode and order.partner_id.zip != order.zipcode:
```

Comment:
Replace `order.partner_id.zip != order.zipcode` with `not order.partner_id.zip` according to the requirement. I'm not sure if the original requirement will be changed later but we can stick to it for now.